### DISCUSSION TITLE
Open Webui  Function call RagFlow API

### DISCUSSION THREAD

**Discussion by ixqkedp** [open-webui/open-webui#5045] [General]
I successfully called the RagFlow knowledge base using the Function Filter of Open Webui. However, in my current implementation, when a user starts a new topic in Open Webui, the conversation_id in my program remains the same. As a result, even though a new topic is started in Open Webui, it is still considered the old topic on the RagFlow side.

**I would like to know if there is any parameter in Open Webui that can notify me when a user starts a new topic, so that I can use the RagFlow API to update the conversation_id (as shown in step 1 of the code below).**

```
from pydantic import BaseModel, Field
from typing import Optional
import requests  # Importing the requests library to handle HTTP requests

from utils.misc import (
    get_last_user_message,
    add_or_update_system_message,
)

class Filter:
    class Valves(BaseModel):
        # Template for modifying context information in the user interface
        RAG_TEMPLATE: str = Field(
            default="""Use the following context as your learned knowledge, inside <context></context> XML tags.
                    <context>
                        {{CONTEXT}}
                    </context>
                    
                    When answering to user:
                    - If you don't know, just say that you don't know.
                    - If you're unsure, ask for clarification.
                    Avoid mentioning that you obtained the information from the context.
                    And answer according to the language of the user's question.
                    
                    Given the context information, answer the user query.""",
            description="Template for system messages that embed contextual knowledge.",
        )

    def __init__(self):
        self.file_handler = True  # Enable custom file handling logic
        self.valves = self.Valves()
        self.conversation_id = None

    def inlet(self, body: dict, __user__: Optional[dict] = None) -> dict:
        messages = body.get("messages", [])
        user_query = get_last_user_message(messages)
        context = None

        # Step 1: Get a new conversation ID
        if not self.conversation_id:
            response = requests.get(
                "http://192.168.0.100:8000/v1/api/new_conversation",
                headers={
                    "Authorization": "Bearer ragflow-Q4ZDkzYjA2Njxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
                },
            )
            response.raise_for_status()
            res = response.json()
            self.conversation_id = res["data"]["id"]

        # Step 2: Retrieve historical conversation records
        history_response = requests.get(
            f"http://192.168.0.100:8000/v1/api/conversation/{self.conversation_id}",
            headers={
                "Authorization": "Bearer ragflow-Q4ZDkzYjA2Njxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
            },
        )
        history_response.raise_for_status()
        res = history_response.json()
        history = res["data"]["message"]

        # Step 3: Get the completion response using the conversation ID
        messages = history + [{"content": user_query, "role": "user"}]
        answer_response = requests.post(
            "http://192.168.0.100:8000/v1/api/completion",
            headers={
                "Authorization": "Bearer ragflow-Q4ZDkzYjA2Njxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
            },
            json={
                "conversation_id": self.conversation_id,
                "messages": messages,
                "quote": False,
                "stream": False,
            },
        )
        answer_response.raise_for_status()
        res = answer_response.json()
        context = res["data"]["answer"]

        # Related system message with retrieved context is injected into the message body
        system_message = self.valves.RAG_TEMPLATE.replace("{{CONTEXT}}", context)
        body["messages"] = add_or_update_system_message(system_message, messages)
        return body
```

**Comment by crelocks** [comment]
did you figure this out @ixqkedp ?

**Comment by Constey** [comment]
would be aswell interested in this :-)

**Comment by qqyy30** [comment]
I am looking for this feature as well.

**Comment by unicorn667** [comment]
any news for this topic?

**Comment by kyuchong** [comment]
Please tell me how to solve it
1. OpenWebUI : 10.0.9.75
2. RAGFlow : 10.0.8.56
- API : ragflow-gwODBhZDYwNTAxZDExZjA4OTZiMDI0Mm
- Chat ID : 0953e6ea500311f08a4c0242ac120006

Where should I set each setting?

I am sharing the current RAGFlow log.

**2025-06-25 14:35:00,043 ERROR 30 404 Not Found: The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.
Traceback (most recent call last): 
File "/ragflow/.venv/lib/python3.10/site-packages/flask/app.py", line 880, in full_dispatch_request 
rv = self.dispatch_request() 
File "/ragflow/.venv/lib/python3.10/site-packages/flask/app.py", line 854, in dispatch_request 
self.raise_routing_exception(req) 
File "/ragflow/.venv/lib/python3.10/site-packages/flask/app.py", line 463, in raise_routing_exception 
raise request.routing_exception # type: ignore[misc] 
File "/ragflow/.venv/lib/python3.10/site-packages/flask/ctx.py", line 362, in match_request 
result = self.url_adapter.match(return_rule=True) # type: ignore 
File "/ragflow/.venv/lib/python3.10/site-packages/werkzeug/routing/map.py", line 629, in match 
raise NotFound() from None
werkzeug.exceptions.NotFound: 404 Not Found: The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.**
2025-06-25 14:35:00,045 INFO 30 10.0.9.90 - - [25/Jun/2025 14:35:00] "GET /api/v1/chats_openai/0953e6ea500311f08a4c0242ac120006/models HTTP/1.1" 200 -
